src/python/pants/core/goals/update_build_files.py

```diff
@@ -234,10 +234,17 @@ async def update_build_files(
         raise ValueError(f"Unrecognized formatter: {update_build_files_subsystem.formatter}")
 
     for request in union_membership[RewrittenBuildFileRequest]:
-        if update_build_files_subsystem.fmt and issubclass(request, chosen_formatter_request_class):
+        if update_build_files_subsystem.fmt and request == chosen_formatter_request_class:
             rewrite_request_classes.append(request)
 
-        if update_build_files_subsystem.fix_safe_deprecations or not issubclass(
+        if update_build_files_subsystem.fix_safe_deprecations and issubclass(
```

Comment:
Oh, this is subtle, and I think it was due to my suggestion from before. Thank you for debugging it, and sorry for the incorrect idea.

What do you think about flipping the `if`s so we can get an if/else chain based on the request class and so they become "obviously" mutually exclusive?

```python
if request == chosen_formatter_request_class:
    if update_build_files_subsystem.fmt:
        rewrite_request_classes.append(request)
else if issubclass(request, DeprecationFixerRequest):
    if update_build_files_subsystem.fix_safe_deprecations:
        rewrite_request_classes.append(request)
else:
    rewrite_request_classes.append(request)
```

I guess the next step in DRYing it might be something like:

```python
if request == chosen_formatter_request_class:
    request_enabled = update_build_files_subsystem.fmt
else if issubclass(request, DeprecationFixerRequest):
    request_enabled = update_build_files_subsystem.fix_safe_deprecations
else:
    request_enabled = True

if request_enabled:
    rewrite_request_classes.append(request)
```
But I dno if that's any better (and, as we've learned, don't trust my ideas without double checking them!).

Reply:
I think both suggestions have a bug in them:

```py
if request == chosen_formatter_request_class:
    if update_build_files_subsystem.fmt:
        rewrite_request_classes.append(request)
else if issubclass(request, DeprecationFixerRequest):
    if update_build_files_subsystem.fix_safe_deprecations:
        rewrite_request_classes.append(request)
else:
    rewrite_request_classes.append(request)
```

If Black is the chosen formatter and `FormatWithRuffRequest` is the `request` being checked here, it would fall through to the final `else` clause and get executed incorrectly.

Reply:
I can't say that the version of the logic added in this PR is the best version, but it's the one that made the most sense in my head.

Reply:
> If Black is the chosen formatter and FormatWithRuffRequest is the request being checked here, it would fall through to the final else clause and get added incorrectly.

Oh! You're right. Too hard for me, it seems 🤯